File: LeetCode135.cpp
#include<vector>
#include<numeric>
#include<iostream>
using namespace std;
class Solution {
public:
    int candy(vector<int>& ratings) {
        int n = ratings.size();
        vector<int> left(n);
        for (int i = 0; i < n; i++) {
            if (i > 0 && ratings[i] > ratings[i - 1]) {
                left[i] = left[i - 1] + 1;
            } else {
                left[i] = 1;
            }
        }
        int right = 0, ret = 0;
        for (int i = n - 1; i >= 0; i--) {
            if (i < n - 1 && ratings[i] > ratings[i + 1]) {
                right++;
            } else {
                right = 1;
            }
            ret += max(left[i], right);
        }
        return ret;
    }
};


int main(){
    vector<int> ratings;
    int n=0,t=0;
    cout << "number of children:";
    cin >> n;
    for (int i = 1; i <= n; i++){
        cin >> t;
        ratings.push_back(t);
    }
    Solution sol;
    cout << sol.candy(ratings);
    return 0;
}
// number of children:3
// 1 0 2
// 5

File: LeetCode135.java
import java.util.Scanner;

public class LeetCode135{
    public static void main(String[] args) {
        LeetCode135 solution = new LeetCode135();
        Scanner scanner = new Scanner(System.in);
        int[] ratings;
        int n = 0;
        System.out.print("number of children:");
        n = scanner.nextInt();
        ratings = new int[n];
        for (int i = 1; i <= n; i++){
            ratings[i-1] = scanner.nextInt();
        }
        System.out.println(solution.candy(ratings));
        scanner.close();
    }

    public int candy(int[] ratings) {
        int n = ratings.length;
        int ret = 1;
        int inc = 1, dec = 0, pre = 1;
        for (int i = 1; i < n; i++) {
            if (ratings[i] >= ratings[i - 1]) {
                dec = 0;
                pre = ratings[i] == ratings[i - 1] ? 1 : pre + 1;
                ret += pre;
                inc = pre;
            } else {
                dec++;
                if (dec == inc) {
                    dec++;
                }
                ret += dec;
                pre = 1;
            }
        }
        return ret;
    }
}